streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java

```diff
@@ -165,14 +173,26 @@ public void closeSuspended(final boolean clean,
      */
     public List<ConsumerRecord<byte[], byte[]>> update(final TopicPartition partition,
                                                        final List<ConsumerRecord<byte[], byte[]>> records) {
+        if (records.isEmpty()) {
+            return Collections.emptyList();
+        }
+
         log.trace("Updating standby replicas of its state store for partition [{}]", partition);
-        final long limit = stateMgr.offsetLimit(partition);
+        long limit = offsetLimits.getOrDefault(partition, Long.MAX_VALUE);
 
         long lastOffset = -1L;
         final List<ConsumerRecord<byte[], byte[]>> restoreRecords = new ArrayList<>(records.size());
         final List<ConsumerRecord<byte[], byte[]>> remainingRecords = new ArrayList<>();
 
         for (final ConsumerRecord<byte[], byte[]> record : records) {
+            // Check if we're unable to process records due to an offset limit (e.g. when our
+            // partition is both a source and a changelog). If we're limited then try to refresh
+            // the offset limit if possible.
+            if (record.offset() >= limit && updateableOffsetLimits.contains(partition)) {
+                updateableOffsetLimits.remove(partition);
+                limit = updateOffsetLimits(partition);
```

Comment:
Should we move `updateableOffsetLimits.remove(partition);` into `updateOffsetLimits()` ?

Reply:
Yes, that makes more sense now that we're not removing it in the if stmt.